### DISCUSSION TITLE
Resuming apps from cached State

### DISCUSSION THREAD

**Discussion by ferologics** [pointfreeco/swift-composable-architecture#508] [Ideas]
Hi, I was curious if anyone has attempted to use `@AppState` with the TCA `State`, or provided a default value for some state properties from `UserDefaults`? 

This seems like the natural place to implement some state caching so that the user won't have to re-fetch data on app start. Resuming the app from cached state could also boost the startup time, as the UI should have some snapshot of previously valid state to display on initial launch.

A simple example of state that can be cached is from a chat input field of a draft message:

```swift
extension ChatInputBarState {
    static let live = Self(text: UserDefaults.standard.string(forKey: "draft_message") ?? "")
}
```

This example is contrived, but it illustrates the general direction in which the discussion could go.

I also had this idea to somehow restore the entire app state from a snapshot taken when app was last closed, avoiding the need to write `UserDefaults` boilerplate for each property.

**Comment by AlexisQapa** [comment]
One other way to do it is to send an action when the app is started. When received, the reducer pull the data needed from defaults and load it into the app. This way you are pretty sure you load the data only once and you can also fetch some more if its missing.

**Comment by lukeredpath** [comment]
You could certainly store small bits of information in the user defaults but I wouldn't recommend trying to store too much. 

A simpler option might just be to use Codable and encode your app state to disk as either JSON or a property list. You could write a high level reducer that performs this persistence as an effect in response to certain app lifecycle events or maybe after every state change (but debounced to handle quick fire state changes). 

With this approach you should think carefully about which parts of your state you want to persist. You may not want to persist certain parts of your state tree and just treat it as ephemeral state that resets when the app relaunches. I'd also think carefully about storing large amounts of data that can be easily reloaded from elsewhere and will likely become outdated quickly (e.g. data loaded from an API). 

Another thing to think about with this is how you handle changes to your app structure and missing data when deciding existing persisted state. You'll likely need to override the protocol methods to handle this.
